<h4>
<pre>
name: Jameson Jolley
Date: wed Oct 30th 2024
Class: cs273
Assignment: HW7 ROTATE, MASKING, Analog Digital 

</pre>
</h4>

---

<h4>
<pre>
1. Use logical and/or arithmetic rotation/shift instructions to divide the 32-bit signed number saved 
in R19, R18, R17, R16 (high byte first) by 2^7.  We assume that R19 contains the most significant byte and 
R16 contains the least significant byte.  Save the result in the same registers.
</pre>
</h4>

In [ ]:
L1: LDI R20, 7

ASR R19
ROR R18
ROR R17
ROR R16

DEC R20
BRNE L1

---

<h4><pre>
2. Find the contents of R20 after each of the following is executed:

LDI R20, 0x8A

CLC

ROR R20

ASR R20
</pre></h4>

In [ ]:

LDI R20, 0x8A // r20 := 0b10001010
CLC           // carry flag get set to 0
ROR R20       // r20 := 0b01000101 , c := 0
ASR R20       // r20 := 0b00100010


r20 = 0x22

---

<h4><pre>
3. Write a program to test if the pattern “1010” exists in an 8-bit input binary number. Please use shift instructions.
</pre></h4>


In [ ]:

LDI r16 0b0000-1010
LDI R18 0b//some number
LDI r20, 7 

LOOP: // start of a loop
    MOV R19 , R18
    ANDi R19, 0x0F // removing the upper half
    LSR R18 //shifting to the right
    cp R19, R16 // check for the correct paddern
    BRNE Found 

    DEC R20
    BRNE LOOP

notFound:
    STS val, 0 // sotre the val false
    ret 

Found:
    STS val, 1 // sotre the val True
    ret

---


<h4><pre>
4. Calculate the step sizes of ADC using the following number of bits with Vref=32V. 

     2-bit: 8
     8-bit: 0.125
     16-bit: 0.0005
 </pre></h4>


In [45]:
Vref=32
# this is the method of calulating the step size 
function stepsize(vref,bits)
    return vref/bits
end

stepsize (generic function with 1 method)

In [46]:

#   2-bit:
stepsize(Vref,2^2)

8.0

In [47]:
#   8-bit:
stepsize(Vref,2^8)

0.125

In [48]:
#   16-bit:
stepsize(Vref,2^16)

0.00048828125

---

<h4> 
<pre>
5. With Vref=25.6V and 10 bit ADC, find Vin for the following outputs: 

    Dout = 215
    Dout = 84
    Dout = 233
 </pre>
</h4>

In [49]:
function Vin(Vref,bits,Dout)
    return (Dout/bits)* Vref  
end

Vin (generic function with 1 method)

In [50]:
#lower bound
Vin(25.6,2^10,215)

5.375

In [51]:
#upper bound
Vin(25.6,2^10,216)

5.4

Dout = 215 on $[5.375,5.4)$   

In [52]:
#lower bound
Vin(25.6,2^10,84)

2.1

In [53]:
#upper bound
Vin(25.6,2^10,85)

2.125

Dout = 84 on $[2.1,2.125)$

In [54]:
#lower bound
Vin(25.6,2^10,233)

5.825

In [55]:
#upper bound
Vin(25.6,2^10,234)

5.8500000000000005

Dout = 233 on $[ 5.825 , 5.8500000000000005)$

---

<h4> 
<pre>
6. For the 10-bit ADC in ATmega328p, find the step size for each of the following Vref:

    25 mV
    4096 mV
    50 V

 </pre></h4>

In [56]:
#we can keep this in mv
stepsize(25,2^10)

0.0244140625

    25 mV = 0.0244140625 mv 

In [57]:
#4096mv is 4.096v
stepsize(4.096,2^10)

0.004

    4096 mV = 0.004 v

In [58]:
# we can keep this the same
stepsize(50,2^10)

0.048828125

  50 V = 0.0488v